### PR TITLE
Address update request

### PR DESCRIPTION
**Pull request by raquelromano-gov**
Connects to https://github.com/department-of-veterans-affairs/vets.gov-team/issues/4460

Sets up the address update request with a dummy "Update" button and hard-coded address object, both of which should be replaced when the address form UI is built.

Outstanding questions:
1. Does it matter whether we issue a PATCH or PUT? (@kreek)
2. The example JSON request body is taken from the vets-api swagger, but it fails validation (returns a 422). @kreek can you double-check the validation code and/or let me know if you see an issue with this example address? We're adding the form fields and FE validation soon and I want to make sure I have the right source of truth for what constitutes a valid address.

### CODE REVIEW

File: src/js/letters/actions/letters.js
Comment (raquelromano-gov) [inline src/js/letters/actions/letters.js]: @kreek This address is lifted from the vets-api address swagger but fails validation (returns a 422)

File: src/js/letters/actions/letters.js
Comment (raquelromano-gov) [inline src/js/letters/actions/letters.js]: @kreek can we arbitrarily choose to PATCH or PUT here? Since we have to provide a valid complete address in the request body, I'm guessing a PUT makes more sense.

Comment (kreek) [reply src/js/letters/actions/letters.js]: @raquelromano-gov yes PUT since even with partial updates we send the complete address.

File: src/js/letters/actions/letters.js
Comment (annekainicUSDS) [inline src/js/letters/actions/letters.js]: Are the only options for `type` `DOMESTIC` and `MILITARY`? Wondering how best we're going to send this field because we're not actually including a question for it. I'm guessing it will be dependent on if they select the military address options in the state dropdown, and then we can automatically update this field based on that. Might need a separate action to do that though? I think this will be easier to do once adding the fields is done, so we can include that later.

File: src/js/letters/actions/letters.js
Comment (annekainicUSDS) [inline src/js/letters/actions/letters.js]: Same question here, we're not including an input for this field, so not sure what this value is going to be based on. Also, I've been referring to this ticket for which fields are required or not (https://github.com/department-of-veterans-affairs/vets.gov-team/issues/4415) and `addressEffectiveDate` isn't mentioned, so I assume it's never required?

**Review by annekainicUSDS — COMMENTED** [review]
LGTM so far! Few questions about the different fields we'll be sending, but I'm assuming this will be easier to review as well once I merge my PR into master and we can rebase that on master.

File: src/js/letters/actions/letters.js
Comment (kreek) [reply src/js/letters/actions/letters.js]: there's an `INTERNATIONAL` type too

File: src/js/letters/actions/letters.js
Comment (kreek) [reply src/js/letters/actions/letters.js]: 👍  I'll set the addressEffectiveDate on the back end

File: src/js/letters/actions/letters.js
Comment (kreek) [reply src/js/letters/actions/letters.js]: @raquelromano-gov I think this is an olive branch issue, without the header `"X-Key-Inflection: camel"` that converts camel to snake case I could reproduce the 422, if I set the header then it goes through.

```
curl -X "PUT" "http://localhost:3000/v0/address" \
     -H "X-Key-Inflection: camel" \
     -H "Authorization: Token token=ZQfL15JDekMLBJyvRutJKzS8dCeWYwEjyrq9XhxK" \
     -H "Content-Type: application/json; charset=utf-8" \
     -d $'{
  "zipCode": "97477",
  "city": "Springfield",
  "addressTwo": "",
  "addressEffectiveDate": "2017-08-07T19:43:59.383Z",
  "zipSuffix": "",
  "countryName": "USA",
  "type": "DOMESTIC",
  "addressOne": "225 5th St",
  "addressThree": "",
  "stateCode": "OR"
}'
```
Tested this on DEV (vs INT) and could update an address. Seems like you can set this [per request](https://github.com/department-of-veterans-affairs/vets-website/blob/5e51527b04d4e3fb0eb66c11b2eff077c2c37202/src/js/edu-benefits/1990/actions/index.js#L159) or in [a config](https://github.com/department-of-veterans-affairs/vets-website/blob/5e51527b04d4e3fb0eb66c11b2eff077c2c37202/src/js/edu-benefits/1990/actions/index.js#L159).

File: src/js/letters/actions/letters.js
Comment (raquelromano-gov) [reply src/js/letters/actions/letters.js]: Now tracked in https://github.com/department-of-veterans-affairs/vets.gov-team/issues/4520

File: src/js/letters/actions/letters.js
Comment (raquelromano-gov) [reply src/js/letters/actions/letters.js]: @kreek Thanks for uncovering this! Looks like it's actually a default header for our `GET` requests but I was overriding those for a `POST`, `PUT`, or `PATCH`, so I will update our common JS to default to this as well: https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/js/common/helpers/api.js#L20

Comment (raquelromano-gov) [reply src/js/letters/actions/letters.js]: @kreek Could use your eyeballs again. Now I'm getting a 500

Relevant  vets-api server log:
```
/Users/romano/dsva/vets-api/lib/evss/pciu_address/service.rb:39:in `block in update_address'
/Users/romano/dsva/vets-api/lib/evss/pciu_address/service.rb:47:in `with_exception_handling'
/Users/romano/dsva/vets-api/lib/evss/pciu_address/service.rb:31:in `update_address'
/Users/romano/dsva/vets-api/app/controllers/v0/addresses_controller.rb:13:in `update'
```
https://github.com/department-of-veterans-affairs/vets-api/blob/master/lib/evss/pciu_address/service.rb#L38

Inspecting `address` here with byebug, I see:
```
{"type"=>"DOMESTIC", "address_effective_date"=>"2017-08-29T05:12:51.300+00:00", "address_one"=>"225 5th St", "city"=>"Springfield", "country_name"=>"USA", "state_code"=>"OR", "zip_code"=>"97477"}
```

Request body from Chrome dev tools:
```
{"zipCode":"97477","city":"Springfield","zipSuffix":"","countryName":"USA","type":"DOMESTIC","addressOne":"225 5th St","stateCode":"OR"}
```

Request headers from dev tools:
```
PUT /v0/address HTTP/1.1
Host: localhost:3000
Connection: keep-alive
Content-Length: 136
Pragma: no-cache
Cache-Control: no-cache
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.101 Safari/537.36
Origin: http://localhost:3001
X-Key-Inflection: camel
Authorization: Token token=seddcHi8K3MiyDuZA3aRohE4n6jRXkRHLJ9dmSw_
Content-Type: application/json; charset=utf-8
Accept: */*
Referer: http://localhost:3001/letters/confirm-address
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.8
```

Can you LMK if anything jumps out as strange to you?

**Comment by ayaleloehr** [comment]
@raquelromano-gov what should happen with this PR?

**Comment by raquelromano-gov** [comment]
hanging onto it for the moment while we work out requirements for the EVSS request!

**Comment by raquelromano-gov** [comment]
@cvalarida if you're picking this up, LMK if you have questions on what's currently there--was begun quite a while ago so a bunch of stuff has changed

File: src/js/letters/reducers/index.js
Comment (cvalarida) [inline src/js/letters/reducers/index.js]: I don't know what's supposed to happen when we get a failure, so this is still on the TODO list, but I'm tracking it [here](https://github.com/department-of-veterans-affairs/vets.gov-team/issues/4858).

File: src/js/letters/reducers/index.js
Comment (cvalarida) [inline src/js/letters/reducers/index.js]: This currently doesn't do anything, but I'm tracking that [here](https://github.com/department-of-veterans-affairs/vets.gov-team/issues/4859).

File: test/letters/containers/AddressSection.unit.spec.jsx
Comment (cvalarida) [inline test/letters/containers/AddressSection.unit.spec.jsx]: I figured we made the helpers; why not use them?

File: test/letters/containers/AddressSection.unit.spec.jsx
Comment (cvalarida) [inline test/letters/containers/AddressSection.unit.spec.jsx]: We may consider moving this out of `schemaform-utils` since it's not _necessarily_ schemaform-specific. I just put it there because...it seemed like the place to put it at the time. :man_shrugging:

File: test/util/schemaform-utils.jsx
Comment (cvalarida) [inline test/util/schemaform-utils.jsx]: More helpful for debugging when the name is more descriptive than the ID.

File: test/letters/components/Address.unit.spec.jsx
Comment (cvalarida) [inline test/letters/components/Address.unit.spec.jsx]: As I mentioned in slack, looking for `#errorable-text-input-12` didn't do the trick. I don't know what the _cause_ is, but when I ran mocha on only this file, it worked, but running `npm run test:unit` failed to find the element--presumably because the auto-incremented id didn't match up...but..._Why??_

File: src/js/letters/actions/letters.js
Comment (raquelromano-gov) [inline src/js/letters/actions/letters.js]: I believe we decided it was a PUT

File: src/js/letters/actions/letters.js
Comment (raquelromano-gov) [inline src/js/letters/actions/letters.js]: Good idea: can you write a ticket for it?

File: test/letters/components/Address.unit.spec.jsx
Comment (raquelromano-gov) [reply test/letters/components/Address.unit.spec.jsx]: Not for this PR, but the 3 street name fields should probably have different names/labels for accessibility reasons

File: src/js/letters/actions/letters.js
Comment (cvalarida) [reply src/js/letters/actions/letters.js]: Already done 🙂 
https://github.com/department-of-veterans-affairs/vets.gov-team/issues/4859

File: test/letters/components/Address.unit.spec.jsx
Comment (cvalarida) [reply test/letters/components/Address.unit.spec.jsx]: Yeah, when I ran across this, I made a ticket for it [here](https://github.com/department-of-veterans-affairs/vets.gov-team/issues/4855).

File: test/letters/containers/AddressSection.unit.spec.jsx
Comment (raquelromano-gov) [reply test/letters/containers/AddressSection.unit.spec.jsx]: Hm...yeah I had never used this for non-rjsf tests!

**Review by cvalarida — APPROVED** [review]
Seems wrong that I can approve this after working on it, but @raquelromano-gov reviewed it and said it was :+1: so I'm going for it!

**Comment by raquelromano-gov** [comment]
I vouch for having reviewed and approved your changes! You are still a good person @cvalarida

**Pull request merged**
